### PR TITLE
Update NodeMapper to handle Trait sourceLocation

### PR DESCRIPTION
**Pull request by JordonPhillips**
This updates NodeMapper so that it can properly set source location on traits that use the AbstractTraitBuilder. Unfortunately the fix implemented here is a hack that looks for that specific case, and the same root cause may bite us elsewhere.

The true root cause is that our [default deserializer is unable to resolve generics](https://github.com/awslabs/smithy/blob/main/smithy-model/src/main/java/software/amazon/smithy/model/node/DefaultNodeDeserializers.java#L434). AbstractTraitBuilder generically defines the return value of the setter `sourceLocation` to `B extends AbstractTraitBuilder`. When we use reflection to look at the return type, all we see is `AbstractTraitBuilder`. We *can* also get the type variable, but actually resolving it will take a decent amount of effort as that functionality isn't built in.

This is related to #864 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by gosar** [comment]
Do you think the serializer (Trait->Node) is handling sourceLocation?

File: smithy-model/src/main/java/software/amazon/smithy/model/node/DefaultNodeDeserializers.java [hunk 1]
```diff
@@ -373,8 +374,8 @@ private static boolean isBeanOrBuilderSetter(Method method, Class<?> type, Strin
                 return false;
             }
 
-            // Must either return the target class itself (like a builder) or void.
-            if (method.getReturnType() != void.class && method.getReturnType() != type) {
```
Comment (gosar) [inline smithy-model/src/main/java/software/amazon/smithy/model/node/DefaultNodeDeserializers.java:377]: What about this?
```
            if (!(method.getReturnType() == void.class || method.getReturnType().isAssignableFrom(type))) {
                return false;
            }
```
Isn't as strict as checking that the return type to be exactly the target, but may be ok? Or at least not specific to sourceLocation.

File: smithy-model/src/main/java/software/amazon/smithy/model/node/DefaultNodeSerializers.java [hunk 2]
```diff
@@ -65,6 +65,7 @@ public Node serialize(ToNode value, Set<Object> serializedObjects, NodeMapper ma
                 return null;
             }
 
+            // TODO: make sure every instance of `toNode` is setting this
```
[the remark this answers is not in the document]
Comment (JordonPhillips) [reply smithy-model/src/main/java/software/amazon/smithy/model/node/DefaultNodeSerializers.java:68]: Anything that implements the `ToNode` interface needs to set source location on the node.

File: smithy-model/src/main/java/software/amazon/smithy/model/node/DefaultNodeDeserializers.java (hunk 1 above)
Comment (JordonPhillips) [reply smithy-model/src/main/java/software/amazon/smithy/model/node/DefaultNodeDeserializers.java:377]: Nice! that is a much more tolerable solution

**Review by gosar — APPROVED** [review]

File: smithy-model/src/main/java/software/amazon/smithy/model/node/DefaultNodeSerializers.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gosar) [reply smithy-model/src/main/java/software/amazon/smithy/model/node/DefaultNodeSerializers.java:68]: The comment does not clarify that you are talking about sourceLocation.

**Pull request merged**